### DISCUSSION TITLE
Well, that was hard.

### DISCUSSION THREAD

**Discussion by msklywenn** [outline/outline#4840] [Self Hosting]
Outline seems great. I have set it up on my own server running Debian Bullseye and using Minio for S3 and Apache as proxy. It looks like everything is working now but wow, was it hard to setup. The documentation is a little bit helpful but it's very surprising that most of the information that saved me is scattered on the internet. A bit disappointing for a knowledge app if you ask me 😜

Anyway, I thought I'd compile a list of things that were difficult and that I think would be nice to add to the documentation.

### Auth
- as everyone, I ran into the auth issue that doesn't support any local scheme even if SMTP is setup and expected e-mail login to be working
- I ended up choosing to auth leveraging **Gitea's OpenID as a provider**. The documentation states that a bunch of environment variables must be setup and that looking at .well-known/openid-configuration should give us the values. That's true but that JSON can be very different from one OpenID provider to another so it would be better if each environment variable was accompanied by the actual member from which we should take the value. 
- I also tried using Mattermost's but it seems broken and I'm not sure if Outline or Mattermost is the culprit. 
- The most difficult part was setting up Apache so that it would **proxy the cookies correctly**. That actually is documented for nginx so I shouldn't have ditched that and instead should have directly tried to map that configuration to Apache mod_proxy. Proper Apache mod_proxy settings would still be nice to have though.
- Also, since gitea's preferred_username is not mandatory, I had to tell everyone to set it otherwise they would get a very unhelpful error message. I tried changing OIDC_USERNAME_CLAIM to "name" or "email" ~~but it doesn't seem to change anything. I looked at the code, it seems correct. The problem is a mystery to me.~~ This has since been fixed.

### Minio, massive pain.
- The examples in the documentation must be for local use because **Minio must be visible from the internet** as the client is doing the calls to it, not the server!
- Once that is done, there is a bunch of settings to get right. I never used S3 and it looks like minio is quite different so it's super confusing. First, the bucket's anonymous settings must have an entry for "public/" so that it's readonly for anyone.
- I added my outline domain, with https prefix, to Minio's Cors Allow Origin setting in the API section
- I added a user with readonly/writeonly/readwrite, generated an access key for it and used that as AWS_ACCESS_KEY_ID and matching secret.
- Two other confusing env variables for outline : I set AWS_REGION to "local", AWS_S3_ACL to "public-read"
- This is the write policy I ended up using for the bucket
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:GetBucketLocation"
            ],
            "Resource": [
                "arn:aws:s3:::outline"
            ]
        },
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:ListBucket"
            ],
            "Resource": [
                "arn:aws:s3:::outline"
            ],
            "Condition": {
                "StringEquals": {
                    "s3:prefix": [
                        "public/"
                    ]
                }
            }
        },
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:GetObject"
            ],
            "Resource": [
                "arn:aws:s3:::outline/public/*"
            ]
        }
    ]
}
```
I'm still not sure if all the settings are actually secure though.

Annnnd that's all I can remember on the top of my head. I'm probably missing a few things and I'll head back here to edit the post if I come to recall. Hopefully this post will be interesting to someone.

Looking forward to be using Outline! It's pretty snappy right now on my quite slow machine.

**Comment by aresstokrat** [comment]
I'm still looking for proper demo site. On the main site all 3 auth-id doesn't fits me. Google search shows 1 third party demo with email sign up that doesn't work ((

Can you please help me ? Open for me for few hours to click and try?

**Comment by unwarkz** [comment]
Hi!

Thanks for posting, it was one of the posts that made me finish self-hosted install after all.
What were the symptoms of minio not working correctly?
I now have just "Upload failed" info in left corner when trying to upload anything

**Reply by msklywenn** [reply]
> Hi!
> 
> Thanks for posting, it was one of the posts that made me finish self-hosted install after all. What were the symptoms of minio not working correctly? I now have just "Upload failed" info in left corner when trying to upload anything

That was one of the hard things to debug. The best thing to do is to look into your web browser developer console. The client, not the server, is doing the talk to the minio server. Make sure it is visible from the web browser too (visible from the internet, not just your local network or just localhost)

**Comment by jochenstu** [comment]
I would second a rework of the docs. Instructions for the Docker setup are inadequate and just don't get you a working installation, without working hard to fill in a LOT of blanks. In the spirit of what it enables (self-hosting), the minio setup should be the default to work out of the box.

**Comment by ciekawy** [comment]
could you share more on minio? in example `docker-compose.yml` / `docker.env` I do not see anything like the policy or public access. I am guessing the access from internet can be done via the `https-portal` - I had no any issues with this part, just configured my domain and it just took some two restarts to make all initialised including let's encrypt.

**Comment by ciekawy** [comment]
as for tha auth - slack integration was super straightforward. I tried google auth in test mode but outline complains about non-google-workspace, not sure why...

**Comment by LukasMueller187** [comment]
As I stated in https://github.com/outline/outline/discussions/5106#discussioncomment-5872378 I also did have issues deploying a local outline test instance. Didn't even get to auth. Even though there is a docker compose available.

**Comment by ReessKennedy** [comment]
The work Thomas (@thomasgriffin) did on this blog post helped me with install: https://thomasgriffin.com/how-to-install-the-outline-knowledge-base-wiki-on-ubuntu/

Especially when it came to getting S3 to work.

**Reply by LukasMueller187** [reply]
I appreciate the hint but the prerequisites already contradict the intention to host outline fully locally (without dns and ssl for testing purposes).

**Reply by tommoor** [reply]
Outline was never designed to run locally without DNS and SSL I'm afraid, it's unlikely to be prioritized in the future either.

**Comment by ciekawy** [comment]
I think the https-portal makes it pretty easy to configure local self
signed ssl

**Comment by jenny787** [comment]
if i use the https portal container with an already running reverse proxy (nginx)...do I point the FQDN entry to port 80?  or 443?  or 3000?  I'm guessing 80?

**Comment by ovizii** [comment]
Would you mind sharing more details about the minio setup?
So far I have minio running and am logged in as admin user.
I have created another personal user, logged in as such and then created an access key which is what outline should use for accessing s3.

I'm now already stuck at:

>  added my outline domain, with https prefix, to Minio's Cors Allow Origin setting in the API section

Where exactly do I do that?

**Comment by routerino** [comment]
> Would you mind sharing more details about the minio setup? So far I have minio running and am logged in as admin user. I have created another personal user, logged in as such and then created an access key which is what outline should use for accessing s3.
> 
> I'm now already stuck at:
> 
> > added my outline domain, with https prefix, to Minio's Cors Allow Origin setting in the API section
> 
> Where exactly do I do that?

Seems like a few unnecessary steps, I certainly did not have to mess around with CORS.

I wrote up an [installation guide over here](https://blog.gurucomputing.com.au/Knowledgebases%20with%20Outline/Installing%20Outline%20Knowledgebase/).

**Reply by ovizii** [reply]
Thanks, I will have a look. I have meanwhile found the CORS settings and it immediately fixed my issues.

**Comment by craftzneko** [comment]
Having spent the last 4 weeks fighting to get outline setup I agree documentation could be better. The worst part of this and the bit I am still struggling with is storage with S3. What an absolute pain, we have specific requirements which are not unusual as a government org, we need to self host, use Azure services. So Amazon S3 was out.

Minio appears to have had a gateway feature to proxy S3 requests to other service like azure, and they recently released an Azure marketplace version of minio and funny enough they got rid of the free gateway version...wonder why. So weeks later I am still struggling to get an s3 storage proxy up and running. One big issue is that outline uses what appear to be a fairly unused part of the S3 api spec, Post Requests. 

My only option left is to built an entire VM expose to the internet and then have direct attached storage because minio requires direct IO so azure app service is out. I pretty much hate S3 at this point. Looking at flexify currently who have an enterprise grade S3 proxy and also do a community edition. They had not heard of POST policy but are having their engineering team take a look. After 4 weeks wasted I still don't have a working self hosted outline instance. Not sure I ever will at this point, looking for other products if anyone has any good suggestions but would prefer if someone knows a good solution to getting outline working with azure blob storage

**Reply by LukasMueller187** [reply]
Sounds like a huge pain, so you can't self-host a VM locally with minio? Only using azure? Unfortunately, I have no clue of azure services.

**Reply by craftzneko** [reply]
It has been a pain that is for sure. Fleixfy just confirmed that they dont support S3 post policy and as its niche wont be. So to my knowledge the ONLY way to use outline is to buy amazon S3 storage services which seems weird to lock in to one vendor or minio, who withdrew an essential feature that made it easy to setup. 

I am tempted to setup a VM inside azure simply because of all the effort i have put in so far, but the work involved to setup a VM is so much more than just being able to host a container in Azure's app service, just makes me feel like what gain vs loss of time.

Shame really this has great potential for us and we would have been happy to pay for support but not even being able to host becuase of the rigid requirements for storage has been a deal breaker. Anyway that's my experience other people may have had much better luck and if your in Amazons farm then happy days!

**Reply by LukasMueller187** [reply]
Don't you have a docker environment available? Minio and a reverse proxy is quite fast to set up this way, I'd say.

But I totally agree, I was annoyed out by the S3 storage as well.

**Reply by craftzneko** [reply]
It's kind of what Azure app Service (AAS) is, which is what we are using to host Outline and most of its requirement (PAAS). The problem with Minio is after they did away with the gateway, you could only use Single drive mode on Filesystem but it requires DIRECT_IO support which as AAS is abstracted, we don't control how storage can be attached the best we can do is a volume mount to CIFS which won't support direct_io. There is another service we can use called Azure Kubernetes (AKS) but again we were hoping for a simple approach rather than spinning up a whole management layer just for the storage of some media files. At the moment as annoying as it is the only hope I have is a VM running Minio and all the associated hassle with that. All this for a knowledge base for 10 staff to use, just struggling to justify it, might end up saying use Sharepoint with all its flaws

**Reply by craftzneko** [reply]
Another user just commented to me on this post https://github.com/outline/outline/discussions/1620#discussioncomment-6683498 that Zenko also do a proxy for S3 so I might give that a try 

I also still have a question with the S3proxy repo about if they support S3 Post policy so if I get that going I will update here!

**Comment by nalcabio-tom** [comment]
> * because **Minio must be visible from the internet** as the client is doing the calls to it, not the server!

This is definitely a choice. I just lost an afternoon to this. We have a handful of other services that use minio via their backends only. We'd hate to have to expose minio to the internet just for outline. I assume there aren't any plans to move the s3 connection work to the backend, right?

**Comment by Rustymage** [comment]
I continue to get this issue. I've looked at all the discussions and issues on it and still cannot get it resolved.

`Failed to load resource: Origin URL is not allowed by Access-Control-Allow-Origin. Status code: 200`

I have outline running as docker and Minio separately as I use it elsewhere. Both publically accessible.

Ideas?
